File: Semester 4/Class15_Computer_Contest_Level_3_Notes_and_Homework_Dec_17/lvl3_hw_sol/disjoint_set_test.py
import sys
sys.setrecursionlimit(1000000)
input = sys.stdin.readline

n, m = map(int, input().split())
p = [i for i in range(n+1)]  # create disjoint sets

def find_set(d):   #find leader for dsu containing d
    if d != p[d]:
        p[d] = find_set(p[d])
    return p[d]
ans = []
for i in range(m):
    u, v = map(int, input().split())
    fu = find_set(u)
    fv = find_set(v)
    if fu != fv:
        p[fu] = fv
        ans.append(i+1)
if len(ans) != n-1:
    print('Disconnected Graph')
else:
    for x in ans:
        print(x)

File: Semester 4/Class15_Computer_Contest_Level_3_Notes_and_Homework_Dec_17/lvl3_hw_sol/disjoint_set_test.cpp
#include <bits/stdc++.h>
using namespace std;
int N, M, p[100002]; vector<int> edge;
int Find(int d){ return d==p[d]? p[d]: p[d]=Find(p[d]); }
int main(){
    scanf("%d %d", &N, &M);
    for(int i=1; i<=N; i++) p[i] = i;
    for(int i=1, u, v; i<=M; i++){
        scanf("%d %d", &u, &v);
        int fu = Find(u), fv = Find(v);
        if(fu != fv){ edge.push_back(i); p[fu] = fv; }
        if(edge.size() == N-1) break;
    }
    if(edge.size() < N-1) printf("Disconnected Graph\n");
    else{
        for(int x: edge)
            printf("%d\n", x);
    }
}